### PR TITLE
DA Library Tool Storage Rules

### PR DESCRIPTION
**Pull request by Dala0**
Adds the ability to: -

1. Install named storage rules as XML into the Pluto interface.
2. Add these rules to item shapes via the item page.
3. Delete storage rules from item shapes via the item page.

All of the above require the user to be in the group called 'Admin'.

Tested on two virtual machines.

@fredex42

### CODE REVIEW

File: gnmlibrarytool/plugin.py [hunk 1]
```diff
@@ -122,6 +121,11 @@ def __init__(self):
         self.plugin_guid = "2b849f0f-0509-4edb-98a0-78da05e60682"
 
     def return_string(self,tagname,*args):
-        return {'guid': self.plugin_guid, 'template': 'gnmlibrarytool/mediaview_storagerules.html'}
+        from models import LibraryStorageRule
+        rules = LibraryStorageRule.objects.order_by("storagerule_name")
+        context = args[1]
+        item = context['item']
+        iid = item.getId()
+        return {'guid': self.plugin_guid, 'template': 'gnmlibrarytool/mediaview_storagerules.html', 'context': {'rules': rules, 'item_id': iid}}
```
Comment (fredex42) [inline gnmlibrarytool/plugin.py]: is the item id not already available in the template context?

File: gnmlibrarytool/templates/gnmlibrarytool/delete_failed.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{% themeextends "base.html" %}
+{% load static from staticfiles %}
+{% block title %}Vidispine Library Management{% endblock %}
+
+{% block body %}
+<br />
+<br />
+The attempt to delete a storage rule from an item failed.
```
Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/delete_failed.html]: bit more context/description would be nice.....

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html [hunk 3]
```diff
@@ -6,8 +6,31 @@
     <span id="mvlibtool_sr_loading">
         <img src="/sitemedia/load.gif" height="64"><span>Loading...</span>
     </span>
+    <form action="/gnmlibrarytool/delete/" method="post">
```
Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: would be nicer to do this with javascript than a form i think

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html [hunk 4]
```diff
@@ -6,8 +6,31 @@
...
+        <input type="hidden" name="itemid" value="{{ item_id }}">
+        {% csrf_token %}
+        <table id="mvlibtool_sr_info">
 
-    <table id="mvlibtool_sr_info">
+        </table>
+    </form>
+    {% for group in user.groups.all %}
```
Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: could we not do user.is_admin?

File: gnmlibrarytool/templates/gnmlibrarytool/rule_failed.html [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{% themeextends "base.html" %}
+{% load static from staticfiles %}
+{% block title %}Vidispine Library Management{% endblock %}
+
+{% block body %}
+<br />
+<br />
+The attempt to add a storage rule to an item failed.
```
Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/rule_failed.html]: more context would be useful

File: gnmlibrarytool/plugin.py (hunk 1 above)
Comment (Dala0) [inline gnmlibrarytool/plugin.py]: Not as far as I know.

File: gnmlibrarytool/templates/gnmlibrarytool/delete_failed.html (hunk 2 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/delete_failed.html]: The error message should be printed below.

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html (hunk 3 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: I am not sure what you mean by this.

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html (hunk 4 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: I do not know. I will check.

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html (hunk 3 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: One of the reasons I have done it as I have is that it needs the {% csrf_token %} to function. This is not accessible via JavaScript. It needs to live in the Django template.

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html (hunk 4 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: I tried it and that does not work.

File: gnmlibrarytool/templates/gnmlibrarytool/rule_failed.html (hunk 5 above)
Comment (Dala0) [inline gnmlibrarytool/templates/gnmlibrarytool/rule_failed.html]: The error message should be printed below.

Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/rule_failed.html]: ah, sorry, i didn't see that. let's see how they look in practise

File: gnmlibrarytool/urls.py [hunk 6]
```diff
@@ -30,4 +31,12 @@
     url(r'^diagram$', login_required(DiagramMainView.as_view())),
     url(r'^endpoint/', include(router.urls)),
     url(r'^storageruleinfo/(\w{2}-\d+)$',StorageRuleInfoView.as_view(), name="libtool_storage_info_view"),
+    url(r'^rules/$', login_required(rule_form), name='rules'),
+    url(r'^add/$', login_required(add_rule)),
+    url(r'^addrule/$', login_required(add_rule_to_item)),
+    url(r'^delete/$', login_required(delete_rule)),
```
Comment (fredex42) [inline gnmlibrarytool/urls.py]: I would prefer cleaner naming here. Something like ^itemrule/(?<itemid>\w{2}-\d+)/add and delete for item override views and ^rule/update/(?<ruleid\d+) etc. for defining views in the models. It is not clear what the difference between add and addrule is.

Comment (Dala0) [inline gnmlibrarytool/urls.py]: I am not sure what you mean by override views. 

I will take out one of the lines in urls.py which is pointless.

What does 'defining views in the models' mean?

'add' is for adding a rule to the model that holds rules for later use in the Portal interface.

'addrule' if for adding one of these rules to a Vidispine item.

**Comment by fredex42** [comment]
I have made a couple of improvements to the display.  But, there are a few things that we discussed before that still need addressing:

- when you save a new rule, it should redirect you back to the list page to see the new rule
- (kinda extra, but i thought it was obvious!) if there are no rules defined, then the rule editor table should show a message to explain that there are none there - take a look at gnmgridintegration
- there should be a link in the admin left-hand menu to go to the rule edit as well as the drop-down menu.  Again, take a look at gnmgridintegration for an example.

**Comment by Dala0** [comment]
I have added the three most recent requests. I have also made the system redirect to the list page when you update and delete rules as well.

**Comment by fredex42** [comment]
OK, but the URL syntax improvements are not complete; it's still not obvious what /add or /edit refer to, please see my comment from yesterday evening

File: gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html (hunk 3 above)
Comment (fredex42) [inline gnmlibrarytool/templates/gnmlibrarytool/mediaview_storagerules.html]: http://stackoverflow.com/questions/36131472/django-csrf-token-with-js

File: gnmlibrarytool/plugin.py (hunk 1 above)
Comment (fredex42) [inline gnmlibrarytool/plugin.py]: documentation in the media.html template says

> There is a lot of context on this page, please use the debug tag if you wish to see it all.

not very useful.

But there are examples in there that it looks like you can do {{ item.getId }} directly in the template, and can also do item.getTitle.  Also some interesting stuff around ACLs; around line 18 or so.

**Pull request merged**